bcipy/language/model/causal.py

```diff
@@ -0,0 +1,340 @@
+from collections import Counter
+import torch
+from typing import List, Tuple
+from transformers import AutoModelForCausalLM, AutoTokenizer
+import itertools
+import heapq
+
+from bcipy.language.main import BACKSPACE_CHAR, SPACE_CHAR
+from bcipy.language.main import LanguageModel, ResponseType
+
+from bcipy.helpers.exceptions import InvalidModelException
+
+from scipy.special import logsumexp
+from scipy.special import softmax
+
+from bcipy.config import LM_PATH
+
+
+class CausalLanguageModel(LanguageModel):
+    """Character language model based on a pre-trained causal model, GPT-2 by default."""
+
+    def __init__(self,
+                 response_type: ResponseType,
+                 symbol_set: List[str],
+                 lang_model_name: str = None,
+                 lm_path: str = None,
+                 lm_device: str = "cpu",
+                 lm_left_context: str = "",
+                 ):
+        """
+        Initialize instance variables and load the language model with given path
+        Args:
+            response_type - SYMBOL only
+            symbol_set    - list of symbol strings
+            lang_model_name    - name of the Hugging Face casual language model to load
+            lm_path     - load fine-tuned model from specified directory
+            lm_device        - device to use for making predictions (cpu, mps, or cuda)
+            lm_left_context  - text to condition start of sentence on
+        """
+        super().__init__(response_type=response_type, symbol_set=symbol_set)
+        self.model = None
+        self.tokenizer = None
+        self.vocab_size = 0
+        self.valid_vocab = []
+        self.vocab = {}
+        self.longest_token = 0
+        self.index_to_word = {}
+        self.index_to_word_lower = {}
+        self.model_name = lang_model_name or self.parameters.get("causal_model_name")
+        self.symbol_set_lower = None
+        self.device = lm_device
+        self.left_context = lm_left_context
+
+        # We optionally load the model from a local directory, but if this is not
+        # specified, we load a Hugging Face model
+        local_model_path = lm_path or self.parameters.get("causal_model_path")
+        self.model_dir = f"{LM_PATH}/{local_model_path}" if local_model_path != "" else self.model_name
+
+        # parameters for search
+        self.beam_width = 8
+        self.batch_size = 8
+
+        # backoff to the previous space
+        self.token_backoff = -1
+
+        self.load()
+
+    def supported_response_types(self) -> List[ResponseType]:
+        return [ResponseType.SYMBOL]
+
+    def _convert_space(self, s: str) -> str:
+        ret = ""
+        for ch in s:
+            if ch == ' ':
+                ret += SPACE_CHAR
+            else:
+                ret += ch
+        return ret
```

Comment:
https://docs.python.org/3/library/stdtypes.html#str.replace

    s.replace(' ', SPACE_CHAR)

I would probably just remove this method, since it's redundant.